docs/design/proposals/adoption/resource-adoption.md

```diff
@@ -0,0 +1,225 @@
+    1. Resources exist and have previously been managed by K8s, but are not being actively reconciled
+3. Migrating architecture across K8s environments
+    1. Resource exist and are actively being managed by another K8s environment
+
+# Proposed Implementation
+
+## Overview
+
+The proposed solution will provide an alternate workflow by which to adopt resources through a separate set of manifests. A new custom resource definition, the `AdoptedResource` CRD, provides a standardised template by which to indicate to ACK that it should attempt to adopt a given resource. 
+The `AdoptedResource` CRD provides a boilerplate to reference an existing AWS resource (either through ARN, name or other unique identifier) and a target ACK type. Each ACK manager will subscribe to the list of adopted resources and will attempt to adopt any resources that match its managed type.
+The manager will describe the referenced AWS resource, typically through a `Describe*` SDK call, and use the returned values to compile a new custom resource - with filled spec and status fields. The manager will then apply this custom resource to the cluster using the metadata applied in the `AdoptedResource` spec.
+
+## `AdoptedResource` Specification
+
+The `AdoptedResource` custom resource definition will allow for specification of an AWS resource unique identifier, a target ACK type (with API group and kind), and target metadata fields.
+The AWS resource unique identifier allows identification through at most one of:
+
+* `arn` - An [ARN](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) (Amazon Resource Name)
+* `name` - A resource name (a string value)
+* `id` - A resource ID (an integer value)
+
+The custom resource definition makes no assumptions about which fields should be present for any given target kind. It is therefore up to the manager to validate the identifier input shape for existence and correctness.
+
+The target ACK type is specified through the following required fields:
+
+* [API group and version](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups-and-versioning)
+* [Resource kind](https://kubernetes.io/docs/reference/kubectl/overview/#resource-types)
+
+A validating webhook for the `AdoptedResource` custom resources will validate the existence of the target type in the cluster and ensure the requested type is able to be adopted. 
+
+The target metadata field is an arbitrary mapping of key-value pairs that will be applied as the metadata for the created custom resource. If a `name` field is specified in this mapping, the manager should use this value as the custom resource `name` . Otherwise, the manager will use the name specified in the metadata of the `AdoptedResource` as the default name for the resource. This same method will also apply to the `namespace` field. All other fields in the metadata mapping are applied to the created resource without validation.
```

Comment:
> If a `name` field is specified in this mapping, the manager should use this value as the custom resource `name` . Otherwise, the manager will use the name specified in the metadata of the `AdoptedResource` as the default name for the resource.

I would advise instead having an `AdoptedResource.Spec.Name` field and `AdoptedResource.Spec.Namespace` field and not used metadata key/value mappings for a 'name' and 'namespace' key.

I realize the reason you need this is because `AdoptedResource` is going to be a CRD and you need to have unique names for AdoptedResource CRs across a Kubernetes namespace, even if you have, say, two AdoptedResources for a resource named "production" and one of those resources is an S3 Bucket and the other is an SNS Topic.

However, I think that `AdoptedResource.Spec.Name` and `AdoptedResource.Spec.Namespace` is the appropriate place to put the Kubernetes name and Kubernetes namespace of the CR to be created, just like the `APIVersion` and `Kind` fields indicate the CR's Kubernetes `APIVersion` and `Kind`. Alternatively, you could do something like this:

```yaml
apiVersion: core.services.k8s.aws/v1alpha1
kind: AdoptedResource
metadata:
  name: adopted-data-bucket
spec:
  kubernetes:
    apiVersion: s3.services.k8s.aws/v1alpha
    kind: Bucket
    name: data-bucket
    namespace: team-a
  aws:
    arn:  arn:aws:s3:::data-bucket
```

And group the Kubernetes-identifying information together and separate from the AWS-identifying information.

Reply:
I don't necessarily see how (in the example above) your `kubernetes` object is any different from my arbitrary metadata key/value mapping.

Say you wanted to attach annotations to this S3 bucket, the `kubernetes` object would now have to support an `annotations` map. Then you wanted to attach a label to the bucket, the object would not have to support a `labels` map. This `kubernetes` object is practically replicating the [`ObjectMeta` type](https://github.com/kubernetes/apimachinery/blob/4f505736214f6a32e70f37c6bae217360fe227b2/pkg/apis/meta/v1/types.go#L110).

Are you suggesting the `kubernetes` object should just be of type `ObjectMeta`? I think that was the direction I was intending when describing this arbitrary map, so that's probably my suggestion now.

Reply:
> I don't necessarily see how (in the example above) your `kubernetes` object is any different from my arbitrary metadata key/value mapping.

One is a strongly-structured set of fields. The other is a bag of key/value strings with no structure or schema.

> Say you wanted to attach annotations to this S3 bucket, the `kubernetes` object would now have to support an `annotations` map. Then you wanted to attach a label to the bucket, the object would not have to support a `labels` map. This `kubernetes` object is practically replicating the [`ObjectMeta` type](https://github.com/kubernetes/apimachinery/blob/4f505736214f6a32e70f37c6bae217360fe227b2/pkg/apis/meta/v1/types.go#L110).
> 
> Are you suggesting the `kubernetes` object should just be of type `ObjectMeta`? I think that was the direction I was intending when describing this arbitrary map, so that's probably my suggestion now.

Sure, you could definitely use the `corev1.ObjectMeta` type as the Go type of the `spec.kubenetes` field, that's a good idea.

The thing is that we need to differentiate between metadata that describes the adopted resource CR itself and metadata that describes the CR that will be created *from* the adopted resource identifiers.

Reply:
> The thing is that we need to differentiate between metadata that describes the adopted resource CR itself and metadata that describes the CR that will be created from the adopted resource identifiers.

So from this discussion it sounds like there should be a field marked `metadata` of type `ObjectMeta` in the `spec` of the `AdoptedResource`. The controller will use this as the source of the created CR's `ObjectMeta`, or if not defined then to default to the `name` and `namespace` of the `AdoptedResource` CR?